mh/pat.py

```diff
@@ -2398,26 +2376,82 @@ def sendNtcLayerHost(self, new_leader, seq):
+                        # TODO: Handover host if inside of quest
+                        self.sendNtcCircleBreak(circle, 0)
+                    else:
+                        self.sendNtcCircleLeave(circle, circle_index, 0)
+                except Exception:
+                    pass
+
+                self.session.leave_circle()
+
+                try:
+                    self.sendNtcCircleListLayerChange(circle, circle_index, 0)
+                except Exception:
+                    pass
+            try:
+                new_host = self.session.try_transfer_city_leadership()
+                if new_host:
+                    self.sendNtcLayerHost(new_host, 0)
+            except Exception as e:
+                self.server.error(
+                    'Unable to transfer city leadership. {}'.format(e)
+                )
         if self.session.layer > 0:
-            ntc_data = pati.lp2_string(self.session.capcom_id)
-            self.server.layer_broadcast(self.session, PatID4.NtcLayerOut,
-                                        ntc_data, 0)
-
-    def finish(self):
-        self.finished = True
+            try:
+                ntc_data = pati.lp2_string(self.session.capcom_id)
+                self.server.layer_broadcast(self.session, PatID4.NtcLayerOut,
+                                            ntc_data, 0)
+            except Exception as e:
+                self.server.error(
+                    'Unable to notify player departure. {}'.format(e)
+                )
+
+    def on_exception(self, e):
+        # type: (Exception) -> None
+        self.server.error(traceback.format_exc())
+        self.send_error("{}: {}".format(type(e).__name__, str(e)))
+
+    def on_finish(self):
         try:
-            super(PatRequestHandler, self).finish()
-        except AttributeError:
-            # Swallow exception caused by a bug?
-            #
-            # File "/usr/lib/python2.7/socket.py", line 286, in close
-            #   self._sock.close()
-            # AttributeError: 'NoneType' object has no attribute 'close'
+            self.notify_layer_departure()
+        except Exception:
             pass
 
+        self.session.layer_end()
+        self.session.disconnect()
+        self.session.delete()
+
+        self.server.del_from_debug(self)
+        self.server.info("Client finished!")
+
+    def on_packet(self, data):
+        if data is None:
+            self.finish()
+            return
+
+        packet_id, data, seq = data
+        self.server.debug(
+            "RECV %s[ID=%08x; Seq=%04x]",
+            PAT_NAMES.get(packet_id, "Packet"),
+            packet_id, seq
+        )
+        self.dispatch(packet_id, data, seq)
+
+    def send_packet(self, packet_id=0, data=b'', seq=0):
+        super(PatRequestHandler, self).send_packet(packet_id, data, seq)
+        self.server.debug(
+            "SEND %s[ID=%08x; Seq=%04x]",
+            PAT_NAMES.get(packet_id, "Packet"),
+            packet_id, seq
+        )
```

Comment:
You aren't logging packet data anymore which could be crucial in some debugging/reverse engineering case. What about:
 - switching the current default log level to INFO
 - use `self.server.info` instead of `self.server.debug`
 - add the `hexdump(data)` of the packet in a `self.server.debug` call?

Reply:
Great idea